### ISSUE TITLE
Infer read layout generically

### ISSUE THREAD

**Issue by uniqueg** [zavolanlab/htsinfer#89]
**Is your feature request related to a problem? Please describe.**

Currently, only 3' adapters are inferred and these are identified via scanning for the presence of a list of known adapters. Devise and implement a solution that generically infers complex read layout patterns, including 5' adapters, UMIs and other linkers, other fixed length/sequence adapters etc.

**Describe the solution you'd like**

TBD

**Describe alternatives you've considered**

TBD

**Additional context**

N/A

**Comment by rohank63** [comment]
Can we add a exhaustive list for the 3' adapters ? 
Regarding the 5' adapters , just a question ( may be it doesn't make sense, as not much background in Bio ) 
Can't we reverse the read and then lookout for 5' adapters that are available ?

**Comment by uniqueg** [comment]
Hi @rohank63, I'm afraid it's not so easy. First of all, there is no _exhaustive_ list of adapters, because, in principle, nothing stops people from using any adapter they like. In practice though, most people will use kits to prepare libraries, and those libraries tend to always use the same 3' adapters. And those ones are already in our list, most likely.

For 5' the situation is probably similar (there are kits, and those kits tend to use a finite set of adapters), but the adapters will likely be completely different. A problem here is that there aren't many kits that use 5' adapters that end up in the sequenced reads at all, so we would first need to explore kits and find libraries that have them.

However, even if we do that, there is basically an infinite space of read layouts of fixed, variable or random sequences that can occur in fixed, variable or random positions. There are UMIs and all sorts of other linkers that may be ligated to read fragments, at the start or end of reads or anywhere in between. It might be a single A at the start, a set of 3-6 random nucleotides at the end, a fixed sequence of reads followed by a random UMI and so on and so on.

So basically, if we want to improve on the read layout side of things, we should come up with a general mechanism of identifying those portions of reads that originate from actual transcripts.

@mzavolan had devised a strategy how this could potentially be done, and there is a number of closed issues labeled `read_layout` that all would need to be implemented for this functionality.
